### PR TITLE
[localize] New msg signature with options object

### PR DESCRIPTION
**Pull request by aomarks**
Changes the lit-localize `msg` signature, the first step in adding support for automatic localize message IDs (https://github.com/Polymer/lit-html/issues/1437).

 Before:

  ```ts
  msg(id: string, template: string|TemplateResult|Function, ...args: any[])
  ```

  After:

  ```ts
  msg(template: string|TemplateResult|Function, options: {id: string: args?: any[]})
  ```

In the next PR, `id` will become optional.

Also fixes the json-schema build issue.

### CODE REVIEW

File: packages/localize/src/program-analysis.ts [hunk 1]
```diff
@@ -86,112 +88,250 @@ function extractMsg(
...
+> {
+  if (node === undefined) {
+    return {result: {}};
+  }
+  if (!ts.isObjectLiteralExpression(node)) {
+    return {
+      error: createDiagnostic(
+        file,
+        node,
+        `Expected second argument to msg() to be an object`
```
Comment (justinfagnani) [inline packages/localize/src/program-analysis.ts:155]: "object" -> "object literal"

File: packages/localize/src/program-analysis.ts [hunk 2]
```diff
@@ -86,112 +88,250 @@ function extractMsg(
...
+    if (ts.isIdentifier(property.name)) {
+      name = property.name.escapedText;
+    } else if (ts.isStringLiteral(property.name)) {
+      name = property.name.text;
+    } else {
+      return {
+        error: createDiagnostic(
+          file,
+          property.name,
+          `Options shorthand object assignment must be identifier or string literal`
```
Comment (justinfagnani) [inline packages/localize/src/program-analysis.ts:184]: Is this message correct? Here would would have a computed property name I think.

File: packages/localize/src/program-analysis.ts [hunk 3]
```diff
@@ -86,112 +88,250 @@ function extractMsg(
...
     };
   }
 
-  if (isLitTemplate(contentsArg)) {
-    if (ts.isNoSubstitutionTemplateLiteral(contentsArg.template)) {
-      // E.g. msg('foo', html`bar <b>baz</b>`)
+  let id: string | undefined = undefined;
+  let args: ts.NodeArray<ts.Expression> | undefined = undefined;
+
+  for (const property of node.properties) {
```
Comment (justinfagnani) [inline packages/localize/src/program-analysis.ts:163]: Does `.properties` include object spread?

File: packages/localize/src/program-analysis.ts [hunk 4]
```diff
@@ -86,112 +88,250 @@ function extractMsg(
...
+          ),
+        };
+      }
+      args = property.initializer.elements;
+    } else {
+      return {
+        error: createDiagnostic(
+          file,
+          property,
+          `Options shorthand object property must be "id" or "args"`
```
Comment (justinfagnani) [inline packages/localize/src/program-analysis.ts:220]: Is "shorthand" right here?

File: packages/localize/src/program-analysis.ts [hunk 5]
```diff
@@ -86,112 +88,250 @@ function extractMsg(
...
+}
+
+/**
+ * Analyze the template argument to a msg call.
+ */
+export function extractTemplate(
+  templateArg: ts.Node,
+  file: ts.SourceFile
+): ResultOrError<
+  Pick<ProgramMessage, 'contents' | 'params' | 'isLitTemplate'> & {
```
Comment (justinfagnani) [inline packages/localize/src/program-analysis.ts:236]: nit: I think it'd be nice to give this type a name

**Review by justinfagnani — APPROVED** [review]

File: packages/localize/src/program-analysis.ts (hunk 1 above)
Comment (aomarks) [reply packages/localize/src/program-analysis.ts:155]: Done.

File: packages/localize/src/program-analysis.ts (hunk 3 above)
Comment (aomarks) [reply packages/localize/src/program-analysis.ts:163]: Yep. Here are all the forms, I added this as a comment too:

```ts
const x = {
  a: 0,     // properties > PropertyAssignment > Identifier
  'b': 0,   // properties > PropertyAssignment > StringLiteral
  ['c']: 0, // properties > PropertyAssignment > ComputedPropertyName
  d,        // properties > ShorthandPropertyAssignment
  ...e,     // properties > SpreadAssignment
};
```

https://astexplorer.net/#/gist/4c8b4f5d6e957befe180b429d58fb2ef/1661e78f3669a9fd66c7aafb35040d38a1191cb1

File: packages/localize/src/program-analysis.ts (hunk 2 above)
Comment (aomarks) [reply packages/localize/src/program-analysis.ts:184]: Oops yeah, fixed. I updated both errors to be more clear too.

File: packages/localize/src/program-analysis.ts (hunk 4 above)
Comment (aomarks) [reply packages/localize/src/program-analysis.ts:220]: Nope, copy/paste error I think. Fixed.

File: packages/localize/src/program-analysis.ts (hunk 5 above)
Comment (aomarks) [reply packages/localize/src/program-analysis.ts:236]: Done (`ExtractedTemplate`).

**Pull request merged**